File: longestIncreasingSubsequence.js
class Solution {
  /**
   * @param {number[]} nums
   * @return {number}
   */
  lengthOfLIS(nums) {
    let dp = [];
    dp.push(nums[0]);
    let lis = 1;

    for (let i = 1; i < nums.length; i++) {
      if (nums[i] > dp[dp.length - 1]) {
        dp.push(nums[i]);
        lis++;
        continue;
      }

      let l = 0;
      let r = dp.length - 1;

      while (l < r) {
        let m = Math.floor((l + r) / 2);
        if (dp[m] < nums[i]) {
          l = m + 1;
        } else {
          r = m;
        }
      }
      dp[l] = nums[i];
    }
    return lis;
  }
}

File: longest_increasing_subsequence.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        sequence = []
        sequence.append(nums[0])
        LIS = 1

        for i in range(1, len(nums)):
            if nums[i] > sequence[-1]:
                sequence.push(nums[i])
                LIS += 1
                continue

            l, r = 0, len(sequence) - 1

            while (l < r):
                mid = l + (r - l) // 2
                if sequence[mid] < nums[i]:
                    l = mid+1
                else:
                    r = mid
            sequence[l] = nums[i]

        return LIS
